main/src/ca/uwaterloo/flix/language/phase/Optimizer.scala

```diff
@@ -41,14 +40,263 @@ object Optimizer extends Phase[SimplifiedAst.Root, SimplifiedAst.Root] {
+      case Expression.MkClosure(lambda, freeVars, tpe, loc) =>
+        val (e, b) = optimizeExp(lambda)
+        (Expression.MkClosure(e.asInstanceOf[Expression.Lambda], freeVars, tpe, loc), b)
+      case Expression.MkClosureRef(ref, freeVars, tpe, loc) =>
+        val (e, b) = optimizeExp(ref)
+        (Expression.MkClosureRef(e.asInstanceOf[Expression.Ref], freeVars, tpe, loc), b)
+      case Expression.ApplyRef(sym, args, tpe, loc) =>
+        val (es, bs) = optimizeExps(args).unzip
+        (Expression.ApplyRef(sym, es, tpe, loc), bs.exists(identity))
+      case Expression.ApplyTail(sym, formals, actuals, tpe, loc) =>
+        val (es, bs) = optimizeExps(actuals).unzip
+        (Expression.ApplyTail(sym, formals, es, tpe, loc), bs.exists(identity))
+      case Expression.ApplyHook(hook, args, tpe, loc) =>
+        val (es, bs) = optimizeExps(args).unzip
+        (Expression.ApplyHook(hook, es, tpe, loc), bs.exists(identity))
+      case Expression.Apply(exp, args, tpe, loc) =>
+        val (e, b) = optimizeExp(exp)
+        val (es, bs) = optimizeExps(args).unzip
+        (Expression.Apply(e, es, tpe, loc), bs.exists(identity) || b)
+      case Expression.Unary(op, exp, tpe, loc) =>
+        val (e, b) = optimizeExp(exp)
+        (Expression.Unary(op, e, tpe, loc), b)
+      case Expression.Binary(op, exp1, exp2, tpe, loc) =>
+        // Elimination of run-time tag checks of singleton-valued enums.
+        val (e1, b1) = optimizeExp(exp1)
+        val (e2, b2) = optimizeExp(exp2)
+        (op, e1, e2) match {
+          case (BinaryOperator.Equal, Expression.Tag(sym, tag, exp, tp, lc), _) =>
+            if (singleCaseEnums.contains(sym) && exp == Expression.Unit) (Expression.True, true)
+            else (Expression.Binary(op, e1, e2, tpe, loc), b1 || b2)
+          case (BinaryOperator.Equal, _, Expression.Tag(sym, tag, exp, tp, lc)) =>
+            if (singleCaseEnums.contains(sym) && exp == Expression.Unit) (Expression.True, true)
+            else (Expression.Binary(op, e1, e2, tpe, loc), b1 || b2)
+          case _ => (Expression.Binary(op, e1, e2, tpe, loc), b1 || b2)
+        }
+      case Expression.IfThenElse(exp1, exp2, exp3, tpe, loc) =>
+        // Elimination of dead branches (e.g. if (true) e1 else e2).
+        val (e1, b1) = optimizeExp(exp1)
+        e1 match {
+          case Expression.True => (optimizeExp(exp2)._1, true)
+          case Expression.False => (optimizeExp(exp3)._1, true)
+          case _ =>
+            val (e2, b2) = optimizeExp(exp2)
+            val (e3, b3) = optimizeExp(exp3)
+            (Expression.IfThenElse(e1, e2, e3, tpe, loc), b1 || b2 || b3)
+        }
+      case Expression.Let(sym, exp1, exp2, tpe, loc) =>
+        val (e1, b1) = optimizeExp(exp1)
+        val (e2, b2) = optimizeExp(exp2)
+        (Expression.Let(sym, e1, e2, tpe, loc), b1 || b2)
+      case Expression.Is(exp, tag, loc) =>
+        val (e, b) = optimizeExp(exp)
+        (Expression.Is(e, tag, loc), b)
```

Comment:
If the enum only has a single tag then there is no need for the `Is` expression. It is common to have:

```
type Person(Str, Int)
```

which is really shorthand for:

```
enum Person
  case Person(...)
```

which causes the compiler to generate tests like `p is Person` even though that is always the case when there is only one variant.

(Same for the wrappers we have around Set, Map, etc.)